File: personal_space.py
"""One remembered profile per local Lens installation; never an approval gate."""
import json
import unicodedata

import workspace


def first_name(value):
    if not isinstance(value, str):
        raise ValueError('Enter a first name, or leave it blank.')
    value = value.strip()
    if len(value) > 60 or any(unicodedata.category(c).startswith('C') for c in value):
        raise ValueError('Use up to 60 characters without control characters.')
    return value


def read_profile():
    path = workspace.DATA / 'profile.json'
    try:
        profile = json.loads(path.read_text())
    except FileNotFoundError:
        return None
    except (OSError, ValueError) as exc:
        raise ValueError('Your local profile could not be read. Your portfolio is still available. Keep a backup before repairing the profile.') from exc
    if (not isinstance(profile, dict) or profile.get('version') != 1
            or not isinstance(profile.get('created_at'), str)
            or not isinstance(profile.get('first_name'), str)):
        raise ValueError('This local profile format is not supported. Your portfolio is still available.')
    first_name(profile['first_name'])
    if 'setup_complete' in profile and type(profile['setup_complete']) is not bool:
        raise ValueError('The setup state could not be read.')
    return dict({key: profile[key] for key in ('version', 'first_name', 'created_at')},
                setup_complete=profile.get('setup_complete', True))


def summary():
    # Local reads only. Research drafts and the investing policy remain separate.
    draft = workspace.read_json(workspace.DATA / 'portfolio.json', None)
    saved = isinstance(draft, dict) and isinstance(draft.get('rows'), list)
    manual = workspace.read_json(workspace.HERE / 'manual.json', {})
    sleeves = manual.get('sleeves', {}) if isinstance(manual, dict) else {}
    return {'portfolio_saved': saved,
            'fund_count': len(draft['rows']) if saved else len(sleeves) if isinstance(sleeves, dict) else 0,
            'saved_at': draft.get('saved_at') if saved else None,
            'manual_available': (workspace.HERE / 'portfolio_operating_manual.md').is_file()}


def api(action, payload):
    # Share the export/workspace lock so backups capture complete profile writes.
    with workspace.LOCK:
        profile = read_profile()
        if action == 'bootstrap':
            return {'ok': True, 'profile': profile, 'summary': summary()}
        if action not in ('create', 'update', 'complete-setup'):
            raise ValueError('Unknown personal space action.')
        name = first_name(payload.get('first_name', ''))
        if action == 'create':
            if profile is not None:
                raise ValueError('A space already exists. Reload to open it.')
            profile = {'version': 1, 'first_name': name, 'created_at': workspace.now(), 'setup_complete': False}
        else:
            if profile is None:
                raise ValueError('Create your space first.')
            if action == 'complete-setup': profile['setup_complete'] = True
            else: profile['first_name'] = name
        workspace.atomic_json(workspace.DATA / 'profile.json', profile)
        return {'ok': True, 'profile': profile, 'summary': summary()}

File: personal-space.js
/* Local profile only. This page never calls a broker or submits orders. */
const spaceNode = id => document.getElementById(id);
async function spaceApi(action, payload = {}) {
  const response = await fetch('/api/space/' + action, {
    method: 'POST', headers: {'Content-Type': 'application/json'}, body: JSON.stringify(payload)
  });
  const data = await response.json();
  if (!response.ok || !data.ok) throw Error(data.error || data.log || 'Could not open your space. Try again.');
  return data;
}
function renderSpace(data, focus = false) {
  const profile = data.profile;
  if(profile && profile.setup_complete === false){location.replace('/setup');return}
  spaceNode('welcome').hidden = !!profile;
  spaceNode('home').hidden = !profile;
  if (profile) {
    spaceNode('home-title').textContent = profile.first_name ? 'Welcome back, ' + profile.first_name + '.' : 'Welcome back.';
    const summary = data.summary;
    spaceNode('portfolio-tag').textContent = summary.portfolio_saved ? 'Saved research portfolio' : 'From your operating manual';
    spaceNode('portfolio-summary').textContent = summary.portfolio_saved
      ? summary.fund_count + ' investment' + (summary.fund_count === 1 ? '' : 's') + ' in your saved portfolio. Continue your research or revisit your X-Ray.'
      : 'Start with the investments in your operating manual. Save your research portfolio to make it your own.';
    const date = summary.saved_at ? new Date(summary.saved_at) : null;
    const validDate = date && !Number.isNaN(date.getTime());
    spaceNode('portfolio-time').hidden = !validDate;
    spaceNode('portfolio-time').textContent = validDate ? 'Saved ' + date.toLocaleString(undefined, {dateStyle:'medium', timeStyle:'short'}) : '';
    spaceNode('manual-link').hidden = !summary.manual_available;
    spaceNode('manual-missing').hidden = summary.manual_available;
  }
  if (focus) spaceNode(profile ? 'home-title' : 'welcome-title').focus();
}
async function loadSpace() {
  spaceNode('space-loading').hidden = false;
  spaceNode('space-error').textContent = '';
  spaceNode('load-recovery').hidden = true;
  try { renderSpace(await spaceApi('bootstrap')); }
  catch (error) {
    spaceNode('space-error').textContent = error.message;
    spaceNode('load-recovery').hidden = false;
  } finally { spaceNode('space-loading').hidden = true; }
}
async function saveSpace(event, action, inputId, buttonId) {
  event.preventDefault();
  const button = spaceNode(buttonId);
  if (button.disabled) return;
  button.disabled = true;
  spaceNode('space-error').textContent = '';
  try {
    renderSpace(await spaceApi(action, {first_name: spaceNode(inputId).value}), action === 'create');
  } catch (error) { spaceNode('space-error').textContent = error.message; }
  finally { button.disabled = false; }
}
spaceNode('create-space').addEventListener('submit', event => saveSpace(event, 'create', 'first-name', 'create-button'));
spaceNode('retry-load').addEventListener('click', loadSpace);
// Preserve bookmarks from when the research workspace lived at the root URL.
if (['#portfolio', '#xray', '#sources'].includes(location.hash)) {
  location.replace(location.hash==='#sources'?'/profile/data/sources':'/portfolio'+location.hash);
} else {
  loadSpace();
}
